File: Data Structures & Algorithms/count-number-of-islands/submission-4.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        r=len(grid)
        c=len(grid[0])
        seen=set()
        numIslands=0
        directions=[(0,1),(1,0),(-1,0),(0,-1)]
        def valid(row,col):
            return 0<=row<r and 0<=col<c and grid[row][col]=='1'
        def dfs(row,col):
            for dx, dy in directions:
                newRow=row+dx
                newCol=col+dy
                if (newRow,newCol) not in seen and valid(newRow,newCol):
                    seen.add((newRow,newCol))
                    dfs(newRow,newCol)
        
        for i in range(r):
            for j in range(c):
                if (i,j) not in seen and valid(i,j):
                    seen.add((i,j))
                    numIslands+=1
                    dfs(i,j)
        return numIslands

File: Data Structures & Algorithms/count-paths/submission-4.java
class Solution {
    int[][] memo;
    public boolean valid(int[][] grid,int row, int col){
        int r=grid.length;
        int c=grid[0].length;
        return ((0<=row)&&(row<r)&&(0<=col)&&(col<c));
    }
    public int dfs(int[][] grid,int m, int n, int row, int col){
        if ((row==m-1)&&(col==n-1)){
            return 1;
        }
        int total=0;
        if (memo[row][col]!=0){
            return memo[row][col];
        }
        int[][] directions=new int[][] {{0,1},{1,0}};
        for(int[] dir:directions){
            int newRow=row+dir[0];
            int newCol=col+dir[1];
            if (valid(grid,newRow,newCol)){
                total+=dfs(grid,m,n,newRow,newCol);
            }
        }
        memo[row][col]=total;
        return memo[row][col];
    }
    public int uniquePaths(int m, int n) {
        memo=new int[m][n];
        int [][] grid= new int[m][n];
        return dfs(grid,m,n,0,0);
    }
}
